rust/src/storage/s3/mod.rs

```diff
@@ -405,80 +453,91 @@ pub trait LockClient: Send + Sync + Debug {
     /// Attempts to acquire lock. If successful, returns the lock.
     /// Otherwise returns [`Option::None`] which is retryable action.
     /// Visit implementation docs for more details.
-    async fn try_acquire_lock(&self) -> Result<Option<LockItem>, StorageError>;
+    async fn try_acquire_lock(&self, data: &str) -> Result<Option<LockItem>, StorageError>;
 
     /// Returns current lock from DynamoDB (if any).
     async fn get_lock(&self) -> Result<Option<LockItem>, StorageError>;
 
+    /// Update data in the upstream lock of the current user still has it.
+    /// The returned lock will have a new `rnv` so it'll increase the lease duration
+    /// as this method is usually called when the work with a lock is extended.
+    async fn update_data(&self, lock: &LockItem) -> Result<LockItem, StorageError>;
+
     /// Releases the given lock if the current user still has it, returning true if the lock was
     /// successfully released, and false if someone else already stole the lock
     async fn release_lock(&self, lock: &LockItem) -> Result<bool, StorageError>;
 }
 
-#[allow(clippy::unnecessary_wraps)]
-fn try_create_lock_client(_region: Region) -> Result<Option<Box<dyn LockClient>>, StorageError> {
-    match std::env::var("AWS_S3_LOCKING_PROVIDER") {
-        Ok(p) if p.to_lowercase() == "dynamodb" => {
-            cfg_if::cfg_if! {
-                if #[cfg(feature = "dynamodb")] {
-                    let client = dynamodb_lock::DynamoDbLockClient::new(
-                        rusoto_dynamodb::DynamoDbClient::new(_region),
-                        dynamodb_lock::Options::default()
-                    );
-                    Ok(Some(Box::new(client)))
-                } else {
-                    Err(StorageError::Generic("dynamodb feature is not enabled".to_string()))
-                }
+const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+
+impl dyn LockClient {
+    async fn rename_with_lock(
+        &self,
+        s3: &S3StorageBackend,
+        src: &str,
+        dst: &str,
+    ) -> Result<(), StorageError> {
+        let mut lock = self.acquire_lock_loop(src, dst).await?;
+
+        if let Some(ref data) = lock.data {
+            let data: LockData = serde_json::from_str(data)
+                .map_err(|_| StorageError::S3Generic("Lock data deserialize error".to_string()))?;
+
+            let mut rename_result = s3.unsafe_rename_obj(&data.source, &data.destination).await;
```

Comment:
hmm, that sounds legit, wdyt of ignoring a error from repair and just log it, or as you say, just recover `already_exists` only so it won't fail the intended rename?

Reply:
yes, in case of repair, the `already_exist` error should be ignored. other possible errors we should ignore are:

* copy failed due to `src` already deleted by concurrent repair from another writer
* delete failed due to `src` already deleted by concurrent repair from another writer